source/common/router/router.cc

```diff
@@ -211,18 +211,24 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Determine if there is a direct response for the request.
-  if (route_->directResponseEntry()) {
-    auto response_code = route_->directResponseEntry()->responseCode();
-    if (response_code >= Http::Code::MultipleChoices && response_code < Http::Code::BadRequest) {
-      config_.stats_.rq_redirect_.inc();
-      Http::Utility::sendRedirect(*callbacks_, route_->directResponseEntry()->newPath(headers),
-                                  response_code);
-      return Http::FilterHeadersStatus::StopIteration;
-    }
+  const auto* direct_response = route_->directResponseEntry();
+  if (direct_response != nullptr) {
     config_.stats_.rq_direct_response_.inc();
-    sendLocalReply(route_->directResponseEntry()->responseCode(),
-                   route_->directResponseEntry()->responseBody(), false);
-    // TODO(brian-pane) support sending response_headers_to_add.
+    Http::Utility::sendLocalReply(
```

Comment:
This is very similar to the existing Filter::sendLocalReply.  Could we extend that instead?

Reply:
Yes, I _think_ I have a way to make that work. Update coming soon.

Reply:
I removed the `overloaded` parameter from `Http::Filter::sendLocalReply` and replaced it with a more general callback hook to modify the headers. That made the couple of existing calls to that method more verbose, but I think it's a net improvement.